odoo/cli/module.py

```diff
@@ -0,0 +1,211 @@
+
+        db_names = config['db_name']
+        if not db_names or len(db_names) > 1:
+            self.parser.error("Please provide a single database in the config file")
+        parsed_args.db_name = db_names[0]
+        if 'modules' in parsed_args and parsed_args.subcommand in ('install', 'upgrade'):
+            if not parsed_args.modules:
+                parsed_args.modules = '*'
+            elif parsed_args.modules != '*':
+                parsed_args.modules = self._get_module_names(parsed_args.modules)
+
+        match parsed_args.subcommand:
+            case 'install':
+                self._install(parsed_args)
+            case 'uninstall':
+                self._uninstall(parsed_args)
+            case 'upgrade':
+                self._upgrade(parsed_args)
+            case 'force-demo':
+                self._force_demo(parsed_args)
+
+    def _get_zip_path(self, path):
+        if (
+            (fullpath := Path(path).resolve())
+            and fullpath.exists()
+            and fullpath.suffix.lower() == '.zip'
+        ):
+            return fullpath
+        return None
+
+    def _get_module_names(self, module_names):
+        """ Get valid module names from disk before starting the Db environment """
+        initialize_sys_path()
+        module_names = set(module_names)
+        if not module_names:
+            return module_names
+        valid_module_names = {
+            module
+            for module in module_names
+            if get_module_path(module)
+            or self._get_zip_path(module)
+        }
+        if not valid_module_names:
+            _logger.warning("No valid module names found")
+        return valid_module_names
+
+    def _get_modules(self, env, module_names, only_outdated=False):
+        """ Get modules from the db.
+
+            Note: the two version fields' names are inverted
+            as stated in a comment on ir.module.module.
+
+                Roses are red and violets are blue.
+                ``latest_version`` is ``installed_version``, true.
+                That's how it goes sometimes, in Odoo.
+                Is it a bug or a feature, it's up to you.
+
+            :param module_names:  name of the modules
+                                  the upgrade command has a default value '*'
+                                  to indicate all installed modules
+            :param only_outdated: filter only modules that have a newer version on disk.
+        """
+        Module = env['ir.module.module']
+        Module.update_list()
+        domain = []
+        if module_names == '*':
+            # `modules` is optional for upgrade
+            domain.append(('state', '=', 'installed'))
+        elif module_names:
+            domain.append(('name', 'in', module_names))
+        else:
+            return Module
+        modules = Module.search(domain)
+        if only_outdated:
+            modules = modules.filtered(lambda x: x.installed_version >= x.latest_version)
```

Comment:
Is `only_outdated` useful?
BTW you cannot directly compare version strings e.g. for python `'3.10' < '3.9'`

Reply:
Yes, it's useful. Typical use case is you clone a DB with customer's addons and you want to update to your newer code.

Needs to use `parse_version` (from tools).